### DISCUSSION TITLE
Partially allow 'is' pattern in expression tree

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/csharplang#4825] [Language Ideas]
Currently 'is' pattern is not allowed in expression meaning you cannot replace `== null` with `is null`.
I suggest to add support for null checks since they can be mapped to `==` for reference types.

**Comment by HaloFour** [comment]
I'm not sure what you mean.  `is` isn't a pattern, it's an operator that can compare an operand against a pattern.  It can be used anywhere that expects a `bool` expression.  The `null` pattern can be used anywhere that expects a pattern, including within a recursive pattern.

Can you provide an example of where you're trying to use `is` or `null`?

**Comment by theunrepentantgeek** [comment]
There's a semantic difference between `is null` and `== null` that can't be handwaved away.

Reference types can override `==` and do something custom for the `== null` case, but `is null` is specified as always doing that exact check and can't be overridden.

At a higher level, a much more important issue is that the expression trees were designed to carry the full semantic load of the expression through to the various query providers (LINQ for SQL, Entity Framework, etc) so they could use all of that information when generating the appropriate query. 

Adding partial support for newer features by lowering those to existing structures (essentially, throwing out some of the semantic information) makes it impossible to update those query providers as the key information they need is being thrown away.

This is a classic example of a short term dirty trick that'll turn around and cause significant increased pain down the track.

**Comment by NN---** [comment]
I understand that there is a difference.
However you can use `??` in expression which is semantically exactly the same as `is not null`.

```cs
using System;
using System.Linq.Expressions;

public class C {
    public void M() {
        Expression<Func<string,string>> f = x => x != null ? x : "a";
        Expression<Func<string,string>> g = x => x ?? "a";
        
        // Expression<Func<string,string>> h = x => x is notnull ? x : "a";
    }
}
```

[Sample](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHABgAIcCA6AGTQDsBHWgUQA8AHBOKKGgD2TKAG5SpHAGZqeSgGFKAb1KV11WTgAslALIAKAJQq1G81179BIgDw4ArPYLkkNcgD4PlAGaUAvJScAd7BAISBTACuGBiUAPxBlCCUAEQQqRIk5hY8fALCTPZO7m4uXpQA5gFJ/qEJiemZZjmULTkA9B2Ulvk2RY7Oru4VABY1wXVJaFCUTEIw0bEJSSlNWeYAvqSbQA===)

**Reply by HaloFour** [reply]
Ah, you meant expression trees specifically.  No patterns are currently supported in expression trees.  There is a discussion about expanding the types of syntax allowed in expression trees here:  https://github.com/dotnet/csharplang/discussions/158

**Reply by NN---** [reply]
Sorry, for misunderstanding.
Upvoted. Thanks.

**Reply by huoyaoyuan** [reply]
Expression tree is more about what raw code you wrote, not waht they will lower to.
